src/YoctoDao/OffChain.hs

```diff
@@ -0,0 +1,195 @@
+import           PlutusTx.Maybe
+import qualified Data.Map               as Map
+import           Data.Text              (Text)
+import           Data.Void              (Void)
+import           Plutus.Contract        as Contract
+import qualified PlutusTx
+import           PlutusTx.IsData
+import           PlutusTx.Prelude       hiding (Semigroup(..), unless)
+import           Ledger                 hiding (singleton)
+import           Ledger.Ada             as Ada
+import           Ledger.Address         as Address
+import           Ledger.Constraints     as Constraints
+import           Ledger.Index           as Index
+import qualified Ledger.Typed.Scripts   as Scripts
+import qualified Ledger.Contexts                   as Validation
+import           Ledger.Value           as Value
+import           Playground.Contract    (printJson, printSchemas, ensureKnownCurrencies, stage, ToSchema, NonEmpty(..) )
+import           Playground.TH          (mkKnownCurrencies, mkSchemaDefinitions, ensureKnownCurrencies)
+import           Playground.Types       (KnownCurrency (..))
+import           Prelude                (Semigroup (..))
+import           Text.Printf            (printf)
+import           GHC.Generics         (Generic)
+import           Data.String          (IsString (..))
+import           Data.Aeson           (ToJSON, FromJSON)
+import           Playground.Contract
+
+import           YoctoDao.Core
+import           YoctoDao.GovToken
+import           YoctoDao.Treasury
+
+{-- We need to create the proposal using the IdMaker NFT so that it will be valid, also ensure a valid proposal..
+data CreateProposal = CreateProposal
+    { proposalStart :: !POSIXTime
+    , scriptUpdate  :: !Bool
+    , scriptAddr    :: Address
+    , spend         :: !Bool
+    , spendValue    :: Value
+    , spendAddr     :: Address
+    , minting       :: !Bool
+    , minted        :: Integer
+    } deriving (Show, Generic, FromJSON, ToJSON)
+
+createProposal :: forall w s. CreateProposal -> Contract w s Text ()
+createProposal cp = do
+    scriptPubKeyHash <- pubKeyHash <$> Contract.ownPubKey
+    let proposal = Proposal
+            { proposalStart = proposalStart cp
+            , scriptUpdate  = scriptUpdate cp
+            , scriptAddr = scriptAddr cp
+            , spend = spend cp
+            , spendValue = spendValue cp
+            , spendAddr = spendAddr cp
+            , minting = minting cp
+            , minted = minted cp
+            , yes = 0
+            , no = 0
+            } --}
+data SpendOwned = SpendOwned
+    { spendAddr  :: !Address
+    , spendVal   :: !Value
+    , spendDatum :: Integer
+    } deriving (Show, Generic, FromJSON, ToJSON)
+
+type YoctoSchema = Endpoint "1-CreateOwnership" Integer .\/
+                   Endpoint "2-SpendOwnedTx" SpendOwned
+        
+createOwnerId :: forall w s. ValidatorHash -> AssetClass -> AssetClass -> AssetClass -> AssetClass -> AssetClass -> Integer -> Contract w s Text ()
```

Comment:
> AssetClass -> AssetClass -> AssetClass -> AssetClass -> AssetClass

Can prob make the semantics clearer by wrapping these types in `newtype`s with different names, or have a `newtype` parameterised by what these types really are.

This is not just for readability, it also gives you some type safety so that it's impossible to randomly screw up the order of these parameters.

Reply:
Most definitely, I already fixed this in my local version and I am making other fixes as well before we go over things in the meeting. Thanks so much Brian! 💯